### ISSUE TITLE
fs2 mapAsync behavior differs from akka streaming

### ISSUE THREAD

**Issue by nikiforo** [typelevel/fs2#2686]
Consider this code:
```
Stream(10, 3, 2, 1, 1)
  .covary[IO]
  .mapAsync(2) { i =>
    val duration = i.seconds
    for {
      _ <- IO(println(s"going to sleep for $duration"))
      _ <- IO.sleep(duration)
      _ <- IO(println(s"woken up after $duration"))
    } yield i
  }
  .evalMap(i => IO(println(s"after mapAsync/parEvalMap $i")))
  .compile
  .drain
```

Expected log:
```
going to sleep for 10 seconds
going to sleep for 3 seconds
woken up after 3 seconds
woken up after 10 seconds
going to sleep for 2 seconds
going to sleep for 1 second
after mapAsync/parEvalMap 10
after mapAsync/parEvalMap 3
woken up after 1 second
woken up after 2 seconds
going to sleep for 1 second
after mapAsync/parEvalMap 2
after mapAsync/parEvalMap 1
woken up after 1 second
after mapAsync/parEvalMap 1 
```

Actual log:
```
going to sleep for 10 seconds
going to sleep for 3 seconds
woken up after 3 seconds
going to sleep for 2 seconds
woken up after 2 seconds
woken up after 10 seconds
after mapAsync/parEvalMap 10
after mapAsync/parEvalMap 3
after mapAsync/parEvalMap 2
going to sleep for 1 second
going to sleep for 1 second
woken up after 1 second
woken up after 1 second
after mapAsync/parEvalMap 1
after mapAsync/parEvalMap 1 
```

The problem:
3rd element (value: 2) was evaluated before 1st one was completed.
This doesn't seem to be serious issue since maxConcurrent limit wasn't exceeded and later 4th element waits until 1st is completed. 
But such behavior differs from similar (mapAsync) combinator available in akka streaming. There 3rd element waits for completion of 1st.
When coming from akka, my assumption was to have the same behaviour when I saw "mapAsync".
Maybe at least pointing this out in docs/comments would be helpful. Maybe it's just 2.5.8, I didn't check 3.X because I didn't plan to migrate now.

In other words:
the issue is whether the guarantee is expressed in terms of emitted elements or evaluated elements
it respects the latter but breaks the former
this problem doesn't exist in parJoin because as soon as an effect it's done, it breaks order
question is whether the semaphore should be decremented when the effect finishes, or when the final results gets emitted

I've just copied the discussion from [discord](https://discord.com/channels/632277896739946517/632310980449402880/900728753594974228)
Found by @kroczix

**Comment by nikiforo** [comment]
v3.2.3 logs:
```
going to sleep for 10 seconds
going to sleep for 3 seconds
woken up after 3 seconds
woken up after 10 seconds
after mapAsync/parEvalMap 10
after mapAsync/parEvalMap 3
going to sleep for 2 seconds
going to sleep for 1 second
woken up after 1 second
woken up after 2 seconds
after mapAsync/parEvalMap 2
going to sleep for 1 second
after mapAsync/parEvalMap 1
woken up after 1 second
after mapAsync/parEvalMap 1
```

The issue has been resolved.
